service/src/main/java/bio/terra/cbas/controllers/RunSetsApiController.java

```diff
@@ -256,67 +221,64 @@ public ResponseEntity<RunSetStateResponse> postRunSet(RunSetRequest request) {
+    // register RunSet
     RunSet runSet;
-
     try {
-      runSet =
-          new RunSet(
-              runSetId,
-              methodVersion,
-              request.getRunSetName(),
-              request.getRunSetDescription(),
-              request.isCallCachingEnabled(),
-              false,
-              CbasRunSetStatus.UNKNOWN,
-              DateUtils.currentTimeInUTC(),
-              DateUtils.currentTimeInUTC(),
-              DateUtils.currentTimeInUTC(),
-              0,
-              0,
-              objectMapper.writeValueAsString(request.getWorkflowInputDefinitions()),
-              objectMapper.writeValueAsString(request.getWorkflowOutputDefinitions()),
-              request.getWdsRecords().getRecordType(),
-              user.getUserSubjectId(),
-              cbasContextConfiguration.getWorkspaceId());
-    } catch (JsonProcessingException e) {
+      runSet = registerRunSet(request, user, methodVersion);
+    } catch (JsonProcessingException | RunSetCreationException e) {
       log.warn("Failed to record run set to database", e);
       return new ResponseEntity<>(
           new RunSetStateResponse()
-              .errors("Failed to record run set to database. Error(s): " + e.getMessage()),
+              .errors("Failed to register submission request. Error(s): " + e.getMessage()),
+          HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    // create mapping between Record IDs to Run IDs to register runs in database
+    Map<String, UUID> recordIdToRunIdMapping =
+        request.getWdsRecords().getRecordIds().stream()
+            .map(recordId -> Map.entry(recordId, UUID.randomUUID()))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // register Runs
+    List<RunStateResponse> runStateResponseList;
+    try {
+      runStateResponseList = registerRunsInRunSet(runSet, recordIdToRunIdMapping);
+    } catch (RunCreationException e) {
+      String errorMsg =
+          "Failed to record runs to database for RunSet %s".formatted(runSet.runSetId());
+      log.error(errorMsg, e);
+
+      // before marking RunSet in Error state, ensure that any Runs that were registered in database
+      // also get marked as in Error state
+      List<Run> runsInRunSet =
+          runDao.getRuns(new RunDao.RunsFilters(runSet.runSetId(), List.of(QUEUED)));
+      runsInRunSet.forEach(
+          run ->
+              runDao.updateRunStatusWithError(
+                  run.runId(), CbasRunStatus.SYSTEM_ERROR, DateUtils.currentTimeInUTC(), errorMsg));
+
+      // mark RunSet in Error state
+      int runsCount = runsInRunSet.size();
+      runSetDao.updateStateAndRunSetDetails(
+          runSet.runSetId(), CbasRunSetStatus.ERROR, runsCount, runsCount, OffsetDateTime.now());
+
+      return new ResponseEntity<>(
+          new RunSetStateResponse()
+              .errors("Failed to register submission request. Error(s): " + e.getMessage()),
```

Comment:
TOL: This might be a good candidate to extract into a `registerRuns` function in the `RunSetsHelper`/`RunSetsService` layer 🤔

Reply:
Yes it indeed is. I have moved this inside `RunSetsService` in the spirit of making controllers do minimum business logic.